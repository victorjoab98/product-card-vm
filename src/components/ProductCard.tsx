import  React,{ createContext } from 'react'
import styles from '../styles/styles.module.css'
import { useProduct } from '../hooks'
import { InitialValuesProductCard, onChangeProductArgs, Product, ProductCardHandlers, ProductContextProps } from '../interfaces';

export const ProductContext = createContext({} as ProductContextProps);
const { Provider } = ProductContext;

export interface Props {
  product: Product,
  // children?: ReactElement | ReactElement[],
  children: ( args: ProductCardHandlers ) => JSX.Element,
  className?: string,
  style?: React.CSSProperties,
  onChange?: ( args: onChangeProductArgs ) => void,
  value?: number,
  initialValues?: InitialValuesProductCard
}

export const ProductCard = ( { product, children, className, style, onChange, value, initialValues } : Props ) => {

  const { counter, increaseBy, reset, maxCount, isMaxCountReached } = useProduct({ product, onChange, value, initialValues })

  return (
    <Provider value={{
        product,
        counter,
        increaseBy,
        maxCount
    }}> 
        <div className={ `${ styles.productCard } ${ className }` } style={style}>
            { children({
                count: counter,
                isMaxCountReached,
                maxCount,
                product,
                increaseBy,
                reset
            }) }
        </div>
    </Provider>
  )
}
