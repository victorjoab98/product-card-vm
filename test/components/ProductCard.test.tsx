import React from "react"
import rederer from "react-test-renderer"
import ProductCard from "../../src/components"
import { product1 } from "../data/products"

const act = rederer.act;

describe('ProductCard', () => { 
    test('debe mostrarse correctamente', () => {
        const wrapper = rederer.create(
            <ProductCard product={product1}>
                {
                    () => <h1>Product Card</h1>
                }
            </ProductCard>
        )

        expect( wrapper.toJSON() ).toMatchSnapshot();
    })

    test('debe incrementar el contador', () => {
        const wrapper = rederer.create(
            <ProductCard product={product1}>
                {
                    ({ count, increaseBy }) => 
                        <>
                            <h1>Product Card</h1>
                            <span>{ count }</span>
                            <button onClick={() => increaseBy(1)}>+1</button>
                        </>
                }
            </ProductCard>
        )

        let tree = wrapper.toJSON();
        expect( tree ).toMatchSnapshot();

        expect( wrapper.toJSON() ).toMatchSnapshot();

        act( ()=> {
            ( tree as any).children[2].props.onClick();
        })

        tree = wrapper.toJSON();

        expect((tree as any).children[1].children[0]).toBe('1')
    })
 })