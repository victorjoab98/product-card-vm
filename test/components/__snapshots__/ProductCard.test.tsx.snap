// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ProductCard debe incrementar el contador 1`] = `
<div
  className="productCard undefined"
>
  <h1>
    Product Card
  </h1>
  <span>
    0
  </span>
  <button
    onClick={[Function]}
  >
    +1
  </button>
</div>
`;

exports[`ProductCard debe incrementar el contador 2`] = `
<div
  className="productCard undefined"
>
  <h1>
    Product Card
  </h1>
  <span>
    0
  </span>
  <button
    onClick={[Function]}
  >
    +1
  </button>
</div>
`;

exports[`ProductCard debe mostrarse correctamente 1`] = `
<div
  className="productCard undefined"
>
  <h1>
    Product Card
  </h1>
</div>
`;
